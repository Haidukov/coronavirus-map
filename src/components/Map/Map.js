import React, { useState, useCallback }  from 'react';
import GoogleMapReact from 'google-map-react';
import Countries from '../../countries.json';
import Marker from '../Marker';
import Card from '../Card';

import styles from './Map.module.css';

const key = process.env.REACT_APP_GOOGLE_MAPS_API_KEY;

const Map = ({ center, zoom, countries }) => {
    const [selectedCountry, setSelectedCountry] = useState(null);

    const onClick = useCallback(() => ({ lat, lng, x, y, event }) => {
        setSelectedCountry(null);
    }, []);

    const onCircleClick = (country, lat, lng) => {
        setTimeout(() => {
            setSelectedCountry({
                country,
                lat,
                lng
            });
        });
    };

    const getConfirmedInfectionsCircleSize = country => {
        return 0.005 * country.mostRecent.confirmed + 10;
    };

    const getDeathsInfectionsCircleSize = country => {
        return 0.005 * country.mostRecent.deaths + 3;
    }

    const renderCountry = (country, index) => {
        if (!Countries[country.name]) return null;
        const { lat, lng } = Countries[country.name];
        const bigCircleSize = getConfirmedInfectionsCircleSize(country);
        const smallCircleSize = getDeathsInfectionsCircleSize(country);
        return (
            <Marker 
                key={index}
                country={country}
                lat={lat} 
                lng={lng}
                bigCircleSize={bigCircleSize}
                smallCircleSize={smallCircleSize}
                onClick={() => onCircleClick(country, lat, lng)}
            />
        );
    };

    return (
        <div className={styles.map}>
            <GoogleMapReact
                bootstrapURLKeys={{ key }}    
                defaultCenter={center}
                defaultZoom={zoom}
                onClick={onClick}
            >
                {countries.map(renderCountry)}
                {selectedCountry && (
                    <Card 
                        lat={selectedCountry.lat} 
                        lng={selectedCountry.lng}
                        country={selectedCountry.country}
                    />
                )}
            </GoogleMapReact>
        </div>
    )
};

Map.defaultProps = {
    center: {
        lat: 60,
        lng: 30
    },
    zoom: 7
};

export default Map;