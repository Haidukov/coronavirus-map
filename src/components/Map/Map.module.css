.map {
    display: block;
    position: relative;
    height: calc(100vh - 70px);
    width: 100%;
}