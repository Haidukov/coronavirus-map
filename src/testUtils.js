import { act } from '@testing-library/react';
import waait from 'waait'; 

export const wait = () => act(() => waait(0));