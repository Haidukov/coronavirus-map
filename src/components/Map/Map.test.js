import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Map from './Map';
import { wait } from '../../testUtils';

const countriesMock = [
    {
        name: 'Thailand',
        mostRecent: {
            confirmed: 441,
            deaths: 1,
            recovered: 42
        }
    },
    {
        name: 'US',
        mostRecent: {
            confirmed: 10000,
            deaths: 1000,
            recovered: 1000
        }
    }
]

const mapPropsMock = {
    countries: countriesMock,
    zoom: 7,
    center: { lat: 45, lng: 45 } 
}

describe('<Marker/>', () => {
    it('should match snapshot', () => {
        const wrapper = render(
            <Map {...mapPropsMock} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should render data', () => {
        const { getAllByTestId } = render((
            <Map {...mapPropsMock} />
        ));
        const markers = getAllByTestId('marker');
        expect(markers.length).toBe(2);
    });

    it('should handle onclick event', async () => {
        const { getAllByTestId, queryByTestId, getByTestId } = render(<Map {...mapPropsMock} />);
        const [marker] = getAllByTestId('marker');
        let card = queryByTestId('card');
        expect(card).toBeNull();
        fireEvent.click(marker);
        await wait();
        card = getByTestId('card');
        expect(card).toBeInTheDocument();
    });
});