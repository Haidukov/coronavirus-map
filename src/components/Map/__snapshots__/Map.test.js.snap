// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Marker/> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="map"
      >
        <div
          data-testid="marker"
        >
          <div
            class="bigCircle"
            data-testid="bigCircle"
            style="position: relative; top: -6.1025px; left: -6.1025px; width: 12.205px; height: 12.205px;"
          >
            <div
              class="smallCircle"
              data-testid="smallCircle"
              style="width: 3.005px; height: 3.005px;"
            />
          </div>
        </div>
        <div
          data-testid="marker"
        >
          <div
            class="bigCircle"
            data-testid="bigCircle"
            style="position: relative; top: -30px; left: -30px; width: 60px; height: 60px;"
          >
            <div
              class="smallCircle"
              data-testid="smallCircle"
              style="width: 8px; height: 8px;"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="map"
    >
      <div
        data-testid="marker"
      >
        <div
          class="bigCircle"
          data-testid="bigCircle"
          style="position: relative; top: -6.1025px; left: -6.1025px; width: 12.205px; height: 12.205px;"
        >
          <div
            class="smallCircle"
            data-testid="smallCircle"
            style="width: 3.005px; height: 3.005px;"
          />
        </div>
      </div>
      <div
        data-testid="marker"
      >
        <div
          class="bigCircle"
          data-testid="bigCircle"
          style="position: relative; top: -30px; left: -30px; width: 60px; height: 60px;"
        >
          <div
            class="smallCircle"
            data-testid="smallCircle"
            style="width: 8px; height: 8px;"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
